docs/ActiveLabel.rst

```diff
@@ -1,135 +1,320 @@
+By convention, the ``ActiveLabel`` module will be associated with a label equal to the module name. For example,
+in the example above, the ``Actor`` ``ActiveLabel`` module is associated with the ``:Actor`` label. You can 
+customize the label(s) which an ``ActiveLabel`` module is associated with using ``self.associated_labels =``. You must also
+include an ``ActiveLabel`` module in an ``ActiveNode`` class if you want the class to respond to the ``ActiveLabel``.
+
+``ActiveLabel`` modules have several parts:
+
 .. code-block:: ruby
 
-  module Destroyable
-    include Neo4j::ActiveLabel
+  module Actor
+    include Neo4j::ActiveLabel # adds ActiveLabel functionality to the Actor module
 
-    follows_label :Destroyed
+    # ``ActiveLabel`` modules can have associations and properties just like ``ActiveNode`` classes
+    property :popularity
+    has_one :out, :friend, type: :FRIEND, model_class: :Person
 
     included do
-      property :destroyed_at, type: DateTime
+      # When a node is retrieved from the database, it is mapped to an ``ActiveNode`` class and a new
+      # instance of that class is created. We'll call this created object obj A.
+      # If obj A's class includes this ``ActiveLabel``, and, additionally, obj A has the label associated
+      # with this ``ActiveLabel``, then this included block will be evaluated
+      # within the context of obj A.
     end
 
     module InstanceMethods
-      def destroy
-        destroyed_at = Time.now
+      # After obj A has been found and initialized, before the included block is evaluated, obj A will
+      # be extended with these InstanceMethods (e.g. obj.extend(InstanceMethods))
 
-        super
+      def act
+        puts "I acted!"
       end
     end
 
     module ClassMethods
-      def destroyed_recently
-        all.where("#{identity}.destroyed_at > ?", 1.week.ago)
+      # Similar to ``ActiveSupport::Concern``, when this ``ActiveLabel`` module is included in an
+      # ``ActiveNode`` class, the class will be extended with these singleton methods (e.g. Person.extend(ClassMethods))
+
+      def actor_popularity_scale
+        puts "5 stars = excellent. 1 star = poor."
       end
     end
   end
 
+``ActiveLabel`` modules only describe functionality that is tied to a label. Actually adding that label to instances of a class
+is a seperate step. If you'd like to add a label to specific instances of a class, you can use standard ``neo4j-core`` methods
+``add_label()`` or ``remove_label``. You can also use special helper methods that ``ActiveLabel`` adds to a class when it is
+included in a class
 
-Creating
---------
+.. code-block:: ruby
+
+  # Initializes a Person with additional Actor label
+  Person.actor.new
+
+  # Creates a Person with additional Actor label
+  Person.actor.create
+
+  # Creates a Person with additional Actor AND Director labels
+  Person.actor.director.create
```

Comment:
I find this api to be more expressive (and concise) than 
```
person = Person.create
person.add_label(:Actor)
```
The above is also two DB queries, where (`Person.actor.create`) could be one DB query.

Reply:
This being said, I'm wondering what happens given the following
```
module Hollywood
  include Neo4j::ActiveLabel

  self.associated_labels = [:Actor, :Director]
  self.associated_labels_matcher = :any
end

class Person
  include Neo4j::ActiveNode
  include Hollywood
end

person = Person.hollywood.create
```
What labels does `person` have? It could be that we, optionally, let the dev define their own label helper methods. Something like
```
module Hollywood
  include Neo4j::ActiveLabel

  self.associated_labels = [:Actor, :Director]
  self.associated_labels_matcher = :any

  class_label_maker :actor, :Actor
  class_label_maker :director, :Director
end

actor = Person.actor.create
director = Person.director.create
```

Below, I added a comment suggesting that `self.associated_labels` accept a two dimensional array instead of `self.associated_labels_matcher`. What would happen if
`self.associated_labels = [[:Actor, :Director], :Animal]`?

I'm thinking that, if `self.associated_labels = []`, then devs must define their own class label makers. Or maybe we could simply always make devs define their own class label makers, and make it an `ActiveNode` method.
```
module InShowbusiness
  include Neo4j::ActiveLabel

  self.associated_labels = [[:Actor, :Director], :Animal]

  class_label_maker :hollywood, [:Actor, :Director]
  class_label_maker :animal, :Animal
end
```

Reply:
@thefliik again you are bringing up something I had on my mind for a very long time. My idea was to do all this natively with ruby without any changes to the DSL or the api methods.
If each Label maps to an ActiveLabel module to create a node with a set of labels you would construct an object of a class which includes all the ActiveLabels you are interested in. When you read a node from the database and the node's combination of Labels does not correspond to any class you would return an object of an anonymous class which has all the corresponding ActiveLabels included. This feels very natural and intuitive and completely in harmony with the ruby language and how neo4j labels work.

Reply:
@klobuczek Interesting. I think it's important for me to say that this proposal only _adds_ to the existing API, and, even if this were implemented, anyone could continue to use the neo4j gem exactly as they have been without changes.

That being said, your suggestion sounds pretty similar to mine, except that you are getting rid of individual ActiveNode classes and simply using one, anonymous class. Are you suggesting this because you think it would be more flexible?

One downside of your approach, is that (and correct me if I'm wrong), but in your approach a specific label (e.g. `:Person`) could only be association with *one* ActiveLabel module, as oppose to my API in which the `:Person` label could have totally different meanings depending on the ActiveNode class of the object. It also _seems_ like my approach would be easier to integrate into existing applications (simply mixin an ActiveLabel module to your existing ActiveNode class). I could also envision many devs, coming from an ActiveRecord background, ignoring the ActiveLabel functionality in the beginning and building out their app just using ActiveNode (ActiveNode is much more similar, conceptually, to ActiveRecord, after all). At some point in the future when they got more comfortable with Neo4j's polymorphism, they could start mixing ActiveLabel modules into their classes.

I guess what I'm trying to say, is that both strategies are basically the same: neo4jrb gets a node from the database and then wraps that node in a class based on the labels the node has. Then, based on which of those labels are determined to be ActiveLabels, the object gets extended with one or more modules. Why toss `ActiveNode` classes from this process?

Reply:
@thefliik Please note the ActiveNode is not a class but a module, so this is already different from active record, but does offer a huge advantage. What I am suggesting is to not be required to define classes including ActiveNode with properties and associations (which would be standard way like today), but allow defining just modules (including ActiveNode or ActiveLabel or similar,) which you can combine to classes, but which could live alone if an unknown combination of labels is retrieved from the database. This is the only case where I would return anonymous class. This would work nicely for creation and retrieval. I have not thought through the promotion (adding a label to existing node) and demotion (removing a label), which does not seem to have a native counterpart in ruby.